packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -13,685 +21,878 @@ use arrow::{
-                ),
-                PartialSharedState::Write(partial) => (
-                    partial.state_proxy.agent_pool().batches(),
-                    partial.state_proxy.message_pool().batches(),
-                    partial.group_indices.clone(), // TODO: Avoid cloning?
-                ),
-            }
+            let pkg_init = serde_json::to_string(&pkg_init).unwrap();
+            let pkg_init = new_js_string(scope, &pkg_init)?;
+            pkg_init_msgs
+                .set_index(scope, i_pkg, pkg_init.into())
+                .ok_or_else(|| {
+                    JSRunnerError::V8(format!(
+                        "Could not set index {i_pkg} on pkg_init_msgs array"
+                    ))
+                })?;
         }
-    })
-}
-
-fn mem_batch_to_js<'m>(
-    mv8: &'m MiniV8,
-    batch_id: &str,
-    mem: mv8::Object<'m>,
-    persisted: Metaversion,
-) -> Result<mv8::Value<'m>> {
-    let batch = mv8.create_object();
-    let batch_id = mv8.create_string(batch_id);
-    batch.set("id", mv8::Value::String(batch_id))?;
-    batch.set("mem", mem)?;
-    batch.set("mem_version", persisted.memory())?;
-    batch.set("batch_version", persisted.batch())?;
-    Ok(mv8::Value::Object(batch))
-}
-
-fn batch_to_js<'m>(
-    mv8: &'m MiniV8,
-    mem: &Memory,
-    persisted: Metaversion,
-) -> Result<mv8::Value<'m>> {
-    // TODO: Is `mem.data.len()` different from `mem.size`? (like Vec capacity vs len?)
-    let arraybuffer = mv8.create_arraybuffer(mem.data.as_ptr(), mem.size);
-    let batch_id = mem.id();
-    mem_batch_to_js(mv8, batch_id, arraybuffer, persisted)
-}
 
-fn state_to_js<'m, 'a, 'b>(
-    mv8: &'m MiniV8,
-    mut agent_batches: impl Iterator<Item = &'a AgentBatch>,
-    mut message_batches: impl Iterator<Item = &'b MessageBatch>,
-) -> Result<(mv8::Value<'m>, mv8::Value<'m>)> {
-    let js_agent_batches = mv8.create_array();
-    let js_message_batches = mv8.create_array();
-
-    for (i_batch, (agent_batch, message_batch)) in agent_batches
-        .by_ref()
-        .zip(message_batches.by_ref())
-        .enumerate()
-    {
-        let agent_batch = batch_to_js(
-            mv8,
-            agent_batch.batch.segment().memory(),
-            agent_batch.batch.segment().persisted_metaversion(),
-        )?;
-        js_agent_batches.set(i_batch as u32, agent_batch)?;
+        let this = v8::Object::new(scope);
+        let args = &[datasets, pkg_init_msgs.into(), pkg_fns.into()];
+        embedded
+            .start_experiment
+            .call(scope, this.into(), args)
+            .ok_or_else(|| JSRunnerError::V8("Could not call start_experiment".to_string()))?;
 
-        let message_batch = batch_to_js(
-            mv8,
-            message_batch.batch.segment().memory(),
-            message_batch.batch.segment().persisted_metaversion(),
-        )?;
-        js_message_batches.set(i_batch as u32, message_batch)?;
+        Ok(ThreadLocalRunner {
```

Comment:
Is it a strong opinion?

I like to use the actual type name because it makes it easier to know what `Self` is without ide, in hover or having to check in what block you are.

Reply:
I think we're inconsistent at the moment so we could try and make a conscious decision. I also don't think we necessarily _**have**_ to follow this, but the Rust docs themselves use `Self` https://doc.rust-lang.org/std/keyword.struct.html

There's also an argument that in the case of generics `Self` is actually disambiguated because it includes the <T>.

My gut is that we should go with `Self` but I'm open to being persuaded

Reply:
[Here's another example from the std](https://doc.rust-lang.org/src/alloc/string.rs.html#382):

```rs
pub const fn new() -> String {
    String { vec: Vec::new() }
}
```
---

In the end do we have to agree on that at all? From the link you sent me 2 days ago:
> authors: don’t feel obliged to make changes where they can either be (a) handled in follow up, or (b) are more about stylistic preference / code golf than impact.

Reply:
Addressed in https://github.com/hashintel/hash/pull/414/commits/859a20d49bcfe39602c0a041b204a47f2d229c25

Reply:
Thank you! 🙂